### PR TITLE
Added support for SNI

### LINKED ISSUE #754: Add support for SNI hostnames
## What would you like to be added

Send the SNI header with TLS connections.

## Why is this needed

Currently when connecting to TLS hosts, only the IP is used. Some sites are behind passthrough proxies that proxy based upon hostname in the SNI header. Without support for SNI, these zookeeper ensembles cannot be reached via the library.

### PR DESCRIPTION
**Pull request by jessesightler-redhat**
Fixes #754

## Why is this needed?

Supports SNI

## Proposed Changes

Add server_hostname to the wrap call

## Does this PR introduce any breaking change?

No

### CODE REVIEW

**Comment by jessesightler-redhat** [comment]
Formatting issues have been resolved. I'm not sure why one of the pypi tests failed with a segfault when all of the others passed.

File: kazoo/client.py [hunk 1]
```diff
@@ -120,6 +120,7 @@ def __init__(
         ca=None,
         use_ssl=False,
         verify_certs=True,
+        check_hostname=False,
```
Comment (StephenSorriaux) [inline kazoo/client.py:123]: Can you please add a description of this new param L185? This is useful for the documentation.

File: kazoo/handlers/utils.py [hunk 2]
```diff
@@ -241,7 +243,7 @@ def create_tcp_connection(
             # verify_mode to CERT_NONE.
             # TODO: Make hostname verification configurable as some users may
```
Comment (StephenSorriaux) [inline kazoo/handlers/utils.py:244]: Can we remove this TODO since your changes make it completed?

File: kazoo/tests/test_utils.py [hunk 3]
```diff
@@ -30,6 +30,24 @@ def test_timeout_arg(self):
                     timeout = call_args[0][1]
                     assert timeout >= 0, "socket timeout must be nonnegative"
 
+    def test_ssl_server_hostname(self):
```
Comment (StephenSorriaux) [inline kazoo/tests/test_utils.py:33]: If possible, I probably would have liked a test around the new `check_hostname` param. What do you think?

File: kazoo/handlers/utils.py [hunk 4]
```diff
@@ -241,7 +243,7 @@ def create_tcp_connection(
...
             # elect to use it.
-            context.check_hostname = False
+            context.check_hostname = check_hostname
             context.verify_mode = (
                 ssl.CERT_REQUIRED if verify_certs else ssl.CERT_NONE
```
Comment (StephenSorriaux) [inline kazoo/handlers/utils.py:248]: Based on the comment L242, it seems we might need to raise a `ValueError` or something if `check_hostname=True` and `verify_certs=False`. Maybe in the client class when checking the given params. What do you think?

**Review by StephenSorriaux — COMMENTED** [review]
Thank you for this PR. Just have some comments here and there. No worries about the segfault, the pypy testing pipeline just triggers it from time to time, it probably relates to the GA infrastructure, but I really never investigated it since I was not able to replicate it locally.

File: kazoo/client.py (hunk 1 above)
Comment (jessesightler-redhat) [reply kazoo/client.py:123]: I have added a line to the description.

File: kazoo/tests/test_utils.py (hunk 3 above)
Comment (jessesightler-redhat) [reply kazoo/tests/test_utils.py:33]: Good idea, test added

File: kazoo/handlers/utils.py (hunk 2 above)
Comment (jessesightler-redhat) [reply kazoo/handlers/utils.py]: Done :)

File: kazoo/handlers/utils.py (hunk 4 above)
Comment (jessesightler-redhat) [reply kazoo/handlers/utils.py:251]: Agreed. I've added an error for this case.

**Review by StephenSorriaux — DISMISSED** [review]
Thanks a lot, that's great.

I will take care of doing a squash of the commits when merging, so that the final commit message follows our guidelines.

I will also wait a little for another maintainer to gives another feedback.

File: kazoo/handlers/utils.py [hunk 5]
```diff
@@ -237,11 +239,14 @@ def create_tcp_connection(
 
             # Load default CA certs
             context.load_default_certs(ssl.Purpose.SERVER_AUTH)
+            if check_hostname and not verify_certs:
+                raise ValueError(
+                    "Error, if check_hostname is specified "
+                    + "verify_certs must be False"
```
Comment (jeffwidman) [inline kazoo/handlers/utils.py:245]: `"Error"` is redundant, and we could probably word this a little more concisely:

```suggestion
                    "verify_certs must be False when check_hostname is specified"
```

Feel free to break across lines if the linter complains...

File: kazoo/handlers/utils.py [hunk 6]
```diff
@@ -237,11 +239,14 @@ def create_tcp_connection(
...
+                )
             # We must set check_hostname to False prior to setting
             # verify_mode to CERT_NONE.
-            # TODO: Make hostname verification configurable as some users may
-            # elect to use it.
```
Comment (jeffwidman) [inline kazoo/handlers/utils.py:243]: ❤️ for solving this!

**Review by jeffwidman — APPROVED** [review]
I didn't review it deeply as I'm not deeply familiar with SNI, but what I saw seems reasonable. 

Thanks for plumbing it through and also including tests.

File: kazoo/handlers/utils.py (hunk 6 above)
Comment (jessesightler-redhat) [reply kazoo/handlers/utils.py:243]: Thanks... I've incorporated your suggestion for the error text as well.

File: kazoo/handlers/utils.py [hunk 7]
```diff
@@ -241,7 +241,8 @@ def create_tcp_connection(
             context.load_default_certs(ssl.Purpose.SERVER_AUTH)
             if check_hostname and not verify_certs:
                 raise ValueError(
-                    "verify_certs must be False when check_hostname is specified"
+                    "verify_certs must be False when"
```
Comment (ceache) [inline kazoo/handlers/utils.py:244]: From the check on line 242, think the intent is to make sure that the hostname validation is *in addition* to the cert validation, correct?
If so, I think the error msg is misleading.
How about something like "verify_certs must also be True when check_hostname is True" ?

Comment (jessesightler-redhat) [reply kazoo/handlers/utils.py]: Agreed. I have tweaked the wording a bit now.

**Review by ceache — APPROVED** [review]
LGTM. Thanks a lot for your contribution!

**Comment by ceache** [comment]
@StephenSorriaux I believe you wanted to handle the squash & merge?

**Review by StephenSorriaux — APPROVED** [review]

**Pull request merged**